app/src/main/java/com/willowtree/vocable/presets/AddToCategoryPickerFragment.kt

```diff
@@ -0,0 +1,162 @@
+package com.willowtree.vocable.presets
+
+import android.os.Bundle
+import android.view.View
+import android.widget.TextView
+import androidx.core.view.isVisible
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModelProviders
+import androidx.navigation.fragment.findNavController
+import androidx.navigation.fragment.navArgs
+import androidx.viewpager2.widget.ViewPager2
+import com.willowtree.vocable.*
+import com.willowtree.vocable.databinding.FragmentAddToCategoryPickerBinding
+import com.willowtree.vocable.presets.adapter.AddToCategoryPickerViewModel
+import com.willowtree.vocable.room.Category
+import com.willowtree.vocable.settings.SettingsActivity
+import com.willowtree.vocable.utils.VocableFragmentStateAdapter
+import kotlin.math.min
+
+class AddToCategoryPickerFragment : BaseFragment<FragmentAddToCategoryPickerBinding>() {
+
+    private val args: AddToCategoryPickerFragmentArgs by navArgs()
+
+    override val bindingInflater: BindingInflater<FragmentAddToCategoryPickerBinding> =
+        FragmentAddToCategoryPickerBinding::inflate
+    private lateinit var addToCategoryPickerViewModel: AddToCategoryPickerViewModel
+
+    private var maxCategories = 1
+    private lateinit var categoriesAdapter: CategoriesPagerAdapter
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val phraseText = args.phraseText
+
+        binding.backButton.action = {
+            findNavController().popBackStack()
+        }
+
+        val numColumns = resources.getInteger(R.integer.custom_category_columns)
+        val numRows = resources.getInteger(R.integer.custom_category_rows)
+        maxCategories = numColumns * numRows
+
+        categoriesAdapter = CategoriesPagerAdapter(childFragmentManager)
+
+        binding.categoryPagerForwardButton.action = {
+            when (val currentPosition = binding.categoryHolder.currentItem) {
+                categoriesAdapter.itemCount - 1 -> {
+                    binding.categoryHolder.setCurrentItem(0, true)
+                }
+                else -> {
+                    binding.categoryHolder.setCurrentItem(currentPosition + 1, true)
+                }
+            }
+        }
```

Comment:
Pretty cool to pass in lambdas into `action`! I think the one downside of this would be readability that we lose because it doesn't often promote encapsulation. It would be more readable if we we set a member val to hold the lambda and summarize its function.

```
val onForwardButtonClick = {
            when (val currentPosition = binding.categoryHolder.currentItem) {
                categoriesAdapter.itemCount - 1 -> {
                    binding.categoryHolder.setCurrentItem(0, true)
                }
                else -> {
                    binding.categoryHolder.setCurrentItem(currentPosition + 1, true)
                }
            }
        }
...

// Then in onViewCreated you can set it
binding.categoryPagerForwardButton.action = onForwardButtonClick
```

This can be done for all the other actions as well. Maybe this is a paradigm your project has already been promoting, if that's the case then I think it's fine. Just jarring for me to have to read thru all of that in 1 method.